File: apps/web/src/features/workbench/api/experiments.ts
/**
 * Experiment API client — the evaluations a gene set has been through.
 */

import { queryOptions } from "@tanstack/react-query";
import type { Experiment } from "@pathfinder/shared";
import { experimentSchema } from "@pathfinder/shared/generated/zod/experimentSchema";
import { z } from "zod";

import { requestJson } from "@/lib/api/http";

const ExperimentListSchema = z.array(experimentSchema);

/** List the experiments this user ran against one gene set, newest first. */
export function listGeneSetExperiments(geneSetId: string): Promise<Experiment[]> {
  return requestJson(
    ExperimentListSchema,
    `/api/v1/gene-sets/${geneSetId}/experiments`,
  );
}

export function geneSetExperimentsOptions(geneSetId: string) {
  return queryOptions({
    queryKey: ["experiments", "by-gene-set", geneSetId] as const,
    queryFn: () => listGeneSetExperiments(geneSetId),
    enabled: geneSetId !== "",
  });
}

File: apps/api/src/pathfinder/transport/http/routers/gene_sets/experiments.py
"""The evaluations a gene set has been through."""

from fastapi import APIRouter

from pathfinder.services.experiment.store import get_experiment_store
from pathfinder.services.experiment.types import Experiment
from pathfinder.transport.http.deps import CurrentUser

from ._shared import gene_set_service

router = APIRouter()


@router.get("/{gene_set_id}/experiments")
async def list_gene_set_experiments(
    gene_set_id: str,
    user_id: CurrentUser,
) -> list[Experiment]:
    """List this user's experiments on one gene set, newest first."""
    await gene_set_service().get_for_user(user_id, gene_set_id)
    return await get_experiment_store().alist_for_gene_set(gene_set_id, user_id)
